activerecord/lib/active_record/attribute_methods/primary_key.rb

```diff
@@ -163,25 +131,25 @@ def get_primary_key(base_name) # :nodoc:
           #
           #   Project.primary_key # => "foo_id"
           def primary_key=(value)
-            @primary_key        = derive_primary_key(value)
+            @primary_key = if value.is_a?(Array)
+              @composite_primary_key = true
+              include CompositePrimaryKey
```

Comment:
Is it a problem that we might include this more than once if people call it twice?

Reply:
No, Ruby takes care of preventing a module from being included twice, there is one edge case however:

```
module M
end

class A
end

class B < A
end

B.include(M)
A.include(M)

B.ancestors # => [B, M, A, M, Object, Kernel, BasicObject]
```

So the module could be included twice in a hierarchy if an only if you define the CPK in the child before defining it in the parent.

It's doable, but I feel like you are looking for problems if somehow you end up in this situation.

Also if this happens, you will potentially experience a slowdown, but the result will still be valid.